pkg/xen-tools/patches-4.15.0/14-qemu-Init-CPU-mask-per-VCPU.patch

```diff
@@ -0,0 +1,181 @@
+ }
+ 
++static inline void cpumask_clear_bit(uint64_t *mask, uint8_t bit)
++{
++    *mask &= ~(1ul << bit);
++}
++
++static inline long cpumask_get_min_bit(uint64_t mask)
++{
++  return __builtin_ffsll(mask) - 1;
++}
++
++static long pick_pcpu(uint64_t *cpumask)
++{
++    long ret = cpumask_get_min_bit(*cpumask);
++    cpumask_clear_bit(cpumask, ret);
++    return ret;
++}
++
+ void qemu_init_vcpu(CPUState *cpu)
+ {
+     MachineState *ms = MACHINE(qdev_get_machine());
+@@ -2019,6 +2036,20 @@ void qemu_init_vcpu(CPUState *cpu)
+     cpu->nr_threads =  ms->smp.threads;
+     cpu->stopped = true;
+     cpu->random_seed = qemu_guest_random_seed_thread_part1();
++    cpu->pinned = ms->cpu_pin;
++    static uint64_t vm_cpumask;
++    uint64_t vcpu_cpumask;
++    if (!vm_cpumask)
++        vm_cpumask = ms->cpumask;
++    if (!cpu->pinned) {
++        //If the CPUs are not pinned, assign the whole CPU mask to the VCPU
++        vcpu_cpumask = vm_cpumask;
++    } else {
++        //If the CPUs are pinned, pick only one CPU for this VCPU
++        vcpu_cpumask = 1ull << pick_pcpu(&vm_cpumask);
++    }
++
++    cpu->cpumask = vcpu_cpumask;
+ 
+     if (!cpu->as) {
+         /* If the target cpu hasn't set up any address spaces itself,
+diff --git a/tools/qemu-xen/softmmu/vl.c b/tools/qemu-xen/softmmu/vl.c
+index c7c8abc..2276a28 100644
+--- a/tools/qemu-xen/softmmu/vl.c
++++ b/tools/qemu-xen/softmmu/vl.c
+@@ -2829,6 +2829,65 @@ static void create_default_memdev(MachineState *ms, const char *path)
+                             &error_fatal);
+ }
+ 
++static inline unsigned get_max_cpu_in_mask(unsigned int cpumask)
++{
++    return sizeof(cpumask) * BITS_PER_BYTE - 1 - __builtin_clz(cpumask);
++}
++
++static inline void cpumask_set_bit(uint64_t *mask, uint8_t bit)
++{
++    *mask |= 1ull << bit ;
++}
++
++//Parse d[[,-]d]* mask (0-2 or 0-2,5,6). CPU starts with 0.
++//Return 0 in case of error, bitmask if ok
++static uint64_t cpumask_parse(const char* cpumask_str)
++{
++    uint64_t cpumask = 0;
++    const char *cur = cpumask_str;
++    bool range = false;
++
++    if (strcmp(cpumask_str, "") == 0)
++        return 0;
++
++    if (cpumask_str[0] == '-') {
++        warn_report("The CPU mask cannot start with -\n");
++        return 0;
++    }
++
++    uint8_t last_set;
++    while (*cur) {
++        unsigned long num;
++        char *end;
++        if (*cur == '-') {
++            cur++;
++            range = true;
++            continue;
++        }
++        if (*cur == ',') {
++            cur++;
++            continue;
++        }
++        num = strtoul(cur, &end, 10);
++        if (end != cur) {
++            if (num == 0 || num > sizeof(cpumask) * BITS_PER_BYTE)
++                return 0;
++            if (range) {
++                range = false;
++                for (int i = last_set + 1; i < num; i++)
++                    cpumask_set_bit(&cpumask, i);
++            }
++            cpumask_set_bit(&cpumask, num);
++            last_set = num;
```

Comment:
last_set is uint8_t and num is unsinged long.... are we safe with num's value?

Reply:
I will add a check for the value in the num variable.